# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/MacroAssembler.cpp
# Commit: 93f3b8aa5bc5
# Full Hash: 93f3b8aa5bc50a2e4d4e61c7ba0e839b681518a2
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-09 16:10:02
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.1: Untie callWithAbi from Frame::tls for Baseline. r=lth
#   
#   We have a tls dependency in callWithABI for wasm because we call builtins via BuiltinThunk.
#   Last one requires that tls to be loaded. In this patch I gradually extend callWithABI interface to
#   pass offset to tls. This allows us to preserve tls and load it by offset in callWithABI and frees us from
# ==============================================================================

diff -r 372f140d5b96 -r 93f3b8aa5bc5 js/src/jit/MacroAssembler.cpp
--- a/js/src/jit/MacroAssembler.cpp	Wed Sep 09 12:34:23 2020 +0000
+++ b/js/src/jit/MacroAssembler.cpp	Wed Sep 09 09:22:39 2020 +0000
@@ -2458,7 +2458,7 @@
   if (compilingWasm) {
     setupWasmABICall();
     passABIArg(src, MoveOp::DOUBLE);
-    callWithABI(callOffset, wasm::SymbolicAddress::ToInt32);
+    callWithABI(callOffset, wasm::SymbolicAddress::ToInt32, mozilla::Nothing());
   } else {
     setupUnalignedABICall(dest);
     passABIArg(src, MoveOp::DOUBLE);
@@ -3121,6 +3121,7 @@
 
 CodeOffset MacroAssembler::callWithABI(wasm::BytecodeOffset bytecode,
                                        wasm::SymbolicAddress imm,
+                                       mozilla::Maybe<int32_t> tlsOffset,
                                        MoveOp::Type result) {
   MOZ_ASSERT(wasm::NeedsBuiltinThunk(imm));
 
@@ -3134,8 +3135,14 @@
   // The TLS register is used in builtin thunks and must be set, by ABI:
   // reload it after passing arguments, which might have used it at spill
   // points when placing arguments.
-  loadWasmTlsRegFromFrame();
-
+
+  if (tlsOffset) {
+    // Account for stackAdjust and Push(WasmTlsReg).
+    *tlsOffset += stackAdjust + sizeof(void*);
+    loadPtr(Address(getStackPointer(), *tlsOffset), WasmTlsReg);
+  } else {
+    loadWasmTlsRegFromFrame();
+  }
   CodeOffset raOffset = call(
       wasm::CallSiteDesc(bytecode.offset(), wasm::CallSite::Symbolic), imm);
 